File: examples/typegraphs/graphql-server.py
# skip:start
from typegraph import typegraph, Policy, t, Graph
from typegraph.graph.params import Cors
from typegraph.runtimes.http import HttpRuntime


# skip:end
@typegraph(
    # skip:next-line
    cors=Cors(allow_origin=["https://metatype.dev", "http://localhost:3000"]),
)
def graphql_server(g: Graph):
    public = Policy.public()

    github = HttpRuntime("https://api.github.com")

    stargazer = t.struct(
        {
            "login": t.string().rename("login"),
            "user": github.get(
                "/users/{user}",
                t.struct({"user": t.string().from_parent("login")}),
                t.struct({"name": t.string().optional()}),
            ),
        }
    )

    g.expose(
        public,
        stargazers=github.get(
            "/repos/metatypedev/metatype/stargazers?per_page=2",
            t.struct({}),
            t.list(stargazer),
        ),
    )

File: examples/typegraphs/graphql-server.ts
// skip:start
import { Policy, t, typegraph } from "@typegraph/sdk/index.ts";
import { HttpRuntime } from "@typegraph/sdk/runtimes/http.ts";

// skip:end

await typegraph(
  {
    name: "graphql-server",
    // skip:next-line
    cors: { allowOrigin: ["https://metatype.dev", "http://localhost:3000"] },
  },
  (g) => {
    const pub = Policy.public();

    const github = new HttpRuntime("https://api.github.com");

    const stargazer = t.struct({
      login: t.string(),
      user: github.get(
        t.struct({ user: t.string().fromParent("login") }),
        t.struct({ name: t.string().optional() }),
        { path: "/users/{user}" },
      ),
    });

    g.expose(
      {
        stargazers: github.get(t.struct({}), t.list(stargazer), {
          path: "/repos/metatypedev/metatype/stargazers?per_page=2",
        }),
      },
      pub,
    );
  },
);
